### PR TITLE
Enhancement: Form validation when vendor Terms & Conditions content is empty

### PR DESCRIPTION
**Pull request by midu-01**
### All Submissions:

* [x] My code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)
* [x] My code satisfies feature requirements
* [x] My code is tested
* [x] My code passes the PHPCS tests
* [x] My code has proper inline documentation
* [ ] I've included related pull request(s) (optional)
* [ ] I've included developer documentation (optional)
* [x] I've added proper labels to this pull request

---

### Changes proposed in this Pull Request:

Previously, vendors could enable the **Terms & Conditions** option without providing any content.  
This resulted in a blank storefront page for customers, which appeared broken and caused confusion.

This pull request adds **backend form validation** to prevent vendors from saving settings when:
- Terms & Conditions are enabled
- But the T&C content field is empty or contains only HTML/whitespace

The validation ensures that invalid state cannot be saved and improves both vendor and customer experience.

---

### Related Pull Request(s)

* N/A

---

### Closes

* Closes https://github.com/getdokan/dokan-pro/issues/5358)

---

### How to test the changes in this Pull Request:

1. Go to Vendor Dashboard → Settings
2. Enable the **Terms & Conditions** checkbox
3. Leave the T&C content field empty
4. Click **Save Settings**
5. Verify that:
   - Settings are not saved
   - An error message is shown asking to add Terms & Conditions content
6. Add valid T&C content and save again
7. Verify settings save successfully and storefront T&C page renders correctly

---

### Changelog entry

**Title:**  
Prevent enabling vendor Terms & Conditions without content

**Description:**  
Previously, enabling Terms & Conditions without content caused a blank storefront page.  
This change adds backend validation to block saving when the content field is empty, ensuring data integrity and preventing broken storefront pages.

---

### Before Changes

- Vendors could enable Terms & Conditions without providing content
- Storefront displayed a blank page for T&C
- Created confusion and reduced trust for customers

---

### After Changes

- Vendors cannot save settings with empty T&C content
- Clear validation message shown in the dashboard
- Storefront no longer renders blank T&C pages

---

### Feature Video (optional)

N/A

---

### PR Self Review Checklist:

* [x] Code follows code style guidelines
* [x] Naming is clear and understandable
* [x] KISS principle is followed
* [x] DRY principle is respected
* [x] Code is readable and maintainable
* [x] No performance issues introduced
* [x] No unnecessary complexity
* [x] Grammar checked

---

### FOR PR REVIEWER ONLY:

* [ ] Correct — Does the change do what it’s supposed to?
* [ ] Secure — All inputs are properly sanitized and validated
* [ ] Readable — Code is easy to understand and maintain
* [ ] Elegant — Fits well within existing architecture and style

## Summary by CodeRabbit

* **Bug Fixes**
  * Enforces Terms & Conditions content as required when T&C is enabled; empty submissions now show a validation error.
  * Cleans and normalizes submitted T&C text (removes invisible/BOM/control characters and trims) before saving to ensure stored terms are consistent and safe.

* **Style**
  * Adds a red asterisk to the Terms & Conditions label to indicate the field is required.

### CODE REVIEW

**Comment by dev-shahed** [comment]
@midu-01 validation not work for if I escape input by space, enter without any char.

File: includes/Dashboard/Templates/Settings.php [hunk 1]
```diff
@@ -479,6 +479,23 @@ private function store_validate() {
...
         }
 
+        $enable_tnc = isset( $_POST['dokan_store_tnc_enable'] ) && 'on' === sanitize_text_field( wp_unslash( $_POST['dokan_store_tnc_enable'] ) );
+
+        if ( $enable_tnc ) {
+            $store_tnc = isset( $_POST['dokan_store_tnc'] ) ? wp_kses_post( wp_unslash( $_POST['dokan_store_tnc'] ) ) : '';
+
+            $store_tnc_clean = wp_strip_all_tags( $store_tnc );
+            $store_tnc_clean = html_entity_decode( $store_tnc_clean, ENT_QUOTES | ENT_HTML5, 'UTF-8' );
+            $store_tnc_clean = preg_replace( '/[\s\x{00A0}\x{200B}\x{FEFF}]+/u', '', $store_tnc_clean );
```
Comment (MdAsifHossainNadim) [inline includes/Dashboard/Templates/Settings.php]: Can we use the php `trim` function here...?

File: includes/Dashboard/Templates/Settings.php [hunk 2]
```diff
@@ -639,6 +656,20 @@ public function insert_settings_info() {
...
             }
 
+            $store_tnc_raw = isset( $_POST['dokan_store_tnc'] ) ? wp_kses_post( wp_unslash( $_POST['dokan_store_tnc'] ) ) : '';
+
+            $store_tnc_processed = '';
+
+            if ( ! empty( $store_tnc_raw ) ) {
+                $store_tnc_clean = wp_strip_all_tags( $store_tnc_raw );
+                $store_tnc_clean = html_entity_decode( $store_tnc_clean, ENT_QUOTES | ENT_HTML5, 'UTF-8' );
+                $store_tnc_clean = preg_replace( '/[\s\x{00A0}\x{200B}\x{FEFF}]+/u', '', $store_tnc_clean );
```
Comment (MdAsifHossainNadim) [inline includes/Dashboard/Templates/Settings.php]: Same goes to here.

**Review by MdAsifHossainNadim — CHANGES_REQUESTED** [review]

**Comment by dev-shahed** [comment]
@midu-01 Please merge to the latest develop to avoid fatal error..

File: includes/Traits/UnicodeCleaner.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * Unicode cleaning helper trait.
+ *
+ * @since DOKAN_SINCE
+ *
+ * @package Dokan
+ */
+
+namespace WeDevs\Dokan\Traits;
+
+trait UnicodeCleaner {
```
Comment (MdAsifHossainNadim) [inline includes/Traits/UnicodeCleaner.php]: Please create class inside the `Utilities` directory also rename this class, e.g. `RichTextValidatorUtil`.

File: includes/Traits/UnicodeCleaner.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+	/**
+	 * Clean unicode related invisible characters from text.
+	 *
+	 * @param string $text Text to clean.
+	 *
+	 * @return string
+	 */
+	private function unicodeReplace( string $text ): string {
```
Comment (MdAsifHossainNadim) [inline includes/Traits/UnicodeCleaner.php]: Can we update this method name like: `replace_richtext_chars`...? And please use `protected` method rather than private binding.

File: includes/Dashboard/Templates/Settings.php [hunk 5]
```diff
@@ -479,6 +481,23 @@ private function store_validate() {
             }
         }
 
+        $enable_tnc = isset( $_POST['dokan_store_tnc_enable'] ) && 'on' === sanitize_text_field( wp_unslash( $_POST['dokan_store_tnc_enable'] ) );
+
+        if ( $enable_tnc ) {
+            $store_tnc = isset( $_POST['dokan_store_tnc'] ) ? wp_kses_post( wp_unslash( $_POST['dokan_store_tnc'] ) ) : '';
+
+            $store_tnc_clean = wp_strip_all_tags( $store_tnc );
```
Comment (MdAsifHossainNadim) [inline includes/Dashboard/Templates/Settings.php]: Please create a method inside `RichTextValidatorUtil`, e.g. `validate_richtext_content` and handle all the validation from that method.

**Review by MdAsifHossainNadim — CHANGES_REQUESTED** [review]

**Comment by MdAsifHossainNadim** [comment]
## Code Review: PR #3082 — Vendor T&C Content Validation

### Issues Found

#### 1. [ERROR] Regex bug — tabs removed despite comment saying otherwise
**Location:** `includes/Utilities/RichTextValidatorUtil.php:55`

The comment says _"Remove control characters except \n and \t"_ but the regex range `\x00-\x09` includes `\x09` (horizontal tab), so tabs **ARE** removed.

```php
// Current (buggy):
$text = preg_replace( '/[\x00-\x09\x0B\x0C\x0E-\x1F\x7F]/u', '', $text ) ?? '';

// Fix — change upper bound to \x08 to preserve tab (\x09):
$text = preg_replace( '/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/u', '', $text ) ?? '';
```

#### 2. [ERROR] Misleading method name — `validate_richtext_content` doesn't validate
**Location:** `includes/Utilities/RichTextValidatorUtil.php:17`

The method strips HTML tags, decodes entities, and removes special chars — it **cleans/sanitizes** content. A _validator_ should return `bool` or throw. The caller uses this to check emptiness after cleaning:

```php
$store_tnc_clean = RichTextValidatorUtil::validate_richtext_content( $store_tnc );
if ( empty( $store_tnc_clean ) ) { ... }
```

**Fix:** Rename to `clean_richtext_content()` (or `sanitize_richtext_content()`). Also rename the filter from `dokan_validate_richtext_content` to `dokan_clean_richtext_content`. Similarly, rename the class itself to `RichTextCleanerUtil` or `RichTextSanitizerUtil`.

#### 3. [ERROR] Missing class-level PHPDoc
**Location:** `includes/Utilities/RichTextValidatorUtil.php:5`

```php
// Current:
class RichTextValidatorUtil {

// Should be:
/**
 * Utility class for cleaning and normalizing rich text content.
 *
 * @since DOKAN_SINCE
 */
class RichTextValidatorUtil {
```

#### 4. [WARNING] Grammar error in validation message
**Location:** `includes/Dashboard/Templates/Settings.php` → `store_validate()`

```php
// Current:
__( 'Please add Terms & Conditions content before save the settings.', 'dokan-lite' )

// Fix:
__( 'Please add Terms & Conditions content before saving the settings.', 'dokan-lite' )
```

#### 5. [WARNING] Required asterisk always visible regardless of checkbox state
**Location:** `templates/settings/store-form.php:203`

The red `*` is always shown on the "TOC Details" label, but the field is only required when the `dokan_store_tnc_enable` checkbox is checked. If the vendor unchecks the checkbox, the asterisk is misleading.

**Fix:** Toggle the asterisk visibility with JavaScript when the checkbox changes, or conditionally render it server-side based on the current `enable_tnc` value.

#### 6. [WARNING] Inline CSS instead of class
**Location:** `templates/settings/store-form.php:203`

```php
// Current:
<span style="color: red;">*</span>

// Better — use an existing class:
<span class="required">*</span>
```

#### 7. [WARNING] Stored content not actually cleaned of BOM/control chars
**Location:** `includes/Dashboard/Templates/Settings.php` → `insert_settings_info()`

The PR description says _"removes invisible/BOM/control characters ... before saving"_ but the implementation stores `$store_tnc_raw` (the original HTML after `wp_kses_post`), not the cleaned version. The cleaning is only used for the emptiness check:

```php
if ( ! empty( $store_tnc_clean ) ) {
    $store_tnc_processed = $store_tnc_raw;  // ← stores uncleaned version
}
```

If the intent is to also clean the stored content, BOM/control-char removal should be applied to the HTML string (without stripping tags). If the intent is only to validate emptiness, the PR description should be corrected.

#### 8. [WARNING] Duplicate `$_POST` processing
**Location:** `includes/Dashboard/Templates/Settings.php` (two locations)

The raw T&C content is read and processed through `wp_kses_post(wp_unslash())` twice — once in `store_validate()` and again in `insert_settings_info()`. The cleaning utility is also called twice on the same content. Consider reading once and reusing.

#### 9. [WARNING] No tests added
No unit tests for:
- `RichTextValidatorUtil::validate_richtext_content()` — edge cases like NBSP-only, BOM-only, zero-width spaces, mixed content
- The validation logic in `store_validate()` — enabling T&C with empty content, with valid content, with whitespace-only

#### 10. [WARNING] Noisy commit history
17 commits for a ~90-line change. Should be squashed into 1-2 meaningful commits before merge.

#### 11. [SUGGESTION] Zero-width spaces not caught
**Location:** `includes/Utilities/RichTextValidatorUtil.php:48-55`

Characters like zero-width space (U+200B), zero-width joiner (U+200D), and word joiner (U+2060) pass through the current regex. Consider adding:

```php
$text = preg_replace( '/[\x{200B}\x{200C}\x{200D}\x{2060}\x{FEFF}]/u', '', $text );
```

---

### PR Checklist

| Item | Status |
|------|--------|
| WordPress coding standards | ⚠️ Mixed indentation in `replace_richtext_chars()` |
| PHPCS passes | ✅ |
| Inline documentation | ❌ Missing class-level PHPDoc |
| Tests | ❌ No tests added |
| Changelog entry | ✅ |
| `DOKAN_SINCE` placeholders | ✅ Present |

---

### Verdict

The core idea is solid — preventing blank T&C pages is a valid UX improvement. However, please address the **3 ERROR items** (regex bug, naming, PHPDoc) and consider the warnings before merge.

**Comment by MdAsifHossainNadim** [comment]
## Code Review (Re-review): PR #3082 — Vendor T&C Content Validation

### Previous Review Status

| Previous Issue | Status |
|---|---|
| Regex bug (tabs removed) | ✅ Fixed |
| Method name `validate_*` → `sanitize_*` | ✅ Fixed |
| Missing class PHPDoc | ✅ Fixed |
| Grammar "before save" | ✅ Fixed |
| Inline CSS `style="color:red"` | ✅ Fixed |
| Zero-width spaces not caught | ✅ Fixed |

---

### New / Remaining Issues

#### 1. [CRITICAL] Stored T&C content loses ALL HTML formatting
**Location:** `includes/Dashboard/Templates/Settings.php` → `insert_settings_info()` ~line 665

```php
$store_tnc_clean = RichTextValidatorUtil::sanitize_richtext_content( $store_tnc_raw );
$store_tnc_clean = trim( $store_tnc_clean );

if ( ! empty( $store_tnc_clean ) ) {
    $store_tnc_processed = $store_tnc_clean;  // ← stores PLAIN TEXT (all HTML stripped!)
}
```

`sanitize_richtext_content()` calls `wp_strip_all_tags()` — this removes ALL HTML. Storing `$store_tnc_clean` means the vendor's bold, italic, links, paragraphs, and all rich text formatting is **destroyed on save**.

**Fix:** Use the cleaned version only for the emptiness check, store the HTML-sanitized version:

```php
if ( ! empty( $store_tnc_clean ) ) {
    $store_tnc_processed = $store_tnc_raw;  // ← store the wp_kses_post() version (safe HTML)
}
```

`$store_tnc_raw` has already been through `wp_kses_post()`, which is the correct sanitization for stored rich text.

#### 2. [WARNING] Required asterisk always visible regardless of checkbox state
**Location:** `templates/settings/store-form.php:203`

The red `*` always shows on the "TOC Details" label, but the field is only required when the T&C checkbox is checked. Misleading UX when unchecked.

#### 3. [WARNING] Class name still `RichTextValidatorUtil`
**Location:** `includes/Utilities/RichTextValidatorUtil.php`

The method was correctly renamed to `sanitize_richtext_content()`, but the class is still called `RichTextValidatorUtil`. For consistency, rename to `RichTextSanitizerUtil` (and update the filename/import).

---

### Verdict

Great progress — most previous issues are fixed. However, **issue #1 is critical** — it will destroy all rich text formatting when vendors save their T&C. The fix is a one-line change: store `$store_tnc_raw` instead of `$store_tnc_clean`.

**Review by MdAsifHossainNadim — APPROVED** [review]
Looks Good ✅

All previous critical and error issues are resolved. The implementation is clean — validation uses sanitized plain text for emptiness check, stores the `wp_kses_post()` HTML version, class/method naming is consistent, regex handles tabs/zero-width chars correctly.

---

### Suggestions (can be ignored)

1. **Required asterisk always visible regardless of checkbox state**
   - **Location:** `templates/settings/store-form.php:203`
   - The red `*` always shows on "TOC Details" label, but the field is only required when the T&C checkbox is checked. Minor UX note — can be addressed separately if needed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: includes/Dashboard/Templates/Settings.php [hunk 6]
```diff
@@ -511,6 +512,22 @@ private function store_validate() {
             }
         }
 
+        $enable_tnc = isset( $_POST['dokan_store_tnc_enable'] ) && 'on' === sanitize_text_field( wp_unslash( $_POST['dokan_store_tnc_enable'] ) );
+
+        if ( $enable_tnc ) {
+            $store_tnc = isset( $_POST['dokan_store_tnc'] ) ? wp_kses_post( wp_unslash( $_POST['dokan_store_tnc'] ) ) : '';
+
+            $store_tnc_clean = RichTextSanitizerUtil::sanitize_richtext_content( $store_tnc );
+            $store_tnc_clean = trim( $store_tnc_clean );
```
Comment (mrabbani) [inline includes/Dashboard/Templates/Settings.php]: Why do we need to trim after sanitization. Trimming should be done from the `sanitize_richtext_content` method

**Review by mrabbani — APPROVED** [review]
